compiler/src/dotty/tools/dotc/transform/patmat/Space.scala

```diff
@@ -894,32 +914,33 @@ class SpaceEngine(using Context) extends SpaceLogic {
       else
         project(OrType(selTyp, constantNullType, soft = false))
 
+    debug.println(s"targetSpace: ${show(targetSpace)}")
+
     // in redundancy check, take guard as false in order to soundly approximate
     val spaces = cases.map { x =>
       val res =
         if (x.guard.isEmpty) project(x.pat)
         else Empty
 
-      debug.println(s"${x.pat.show} ====> ${res}")
+      debug.println(s"${x.pat.show} ====> ${show(res)}")
       res
     }
 
     (1 until cases.length).foreach { i =>
       val pat = cases(i).pat
-
-      if (pat != EmptyTree) { // rethrow case of catch uses EmptyTree
-        val prevs = Or(spaces.take(i))
-        val curr = project(pat)
+      val prevs = Or(spaces.take(i))
+      if (pat != EmptyTree // rethrow case of catch uses EmptyTree
+        && simplify(intersect(prevs, targetSpace)) != Empty
+        // it's required that at one of the previous cases is reachable (its intersected Space isn't Empty)
+        // because if all the previous cases are unreachable then case i can't be unreachable
+      ) {
```

Comment:
Oh, this isn't a performance improvement, this is correctness!

As explored with Seth there's two levels of unreachable: there's "this case can never be reached for this scrutinee" and there's "this case is never reached, because it's covered by something previous".  We're only doing the second, but in order to do that properly we need to make sure that the previous, including the first(!), is actually possibly reached.  Personally I think the level should be covered by typing rather than here.  But it's also possible, with aliases and casting, for passing code to throw warnings - so it's a bit of a judgement call.

Anyways, rethinking this code, I can make this cheaper by reusing and accumulating, so I avoid re-doing lots of intersecting and simplifying.